metafetcher/fetcher.go

```diff
@@ -0,0 +1,136 @@
+package metafetcher
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/NebulousLabs/skynet-accounts/database"
+
+	"github.com/sirupsen/logrus"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+// maxAttempts defines the maximum number of attempts we are going to make to
+// process a message.
+const maxAttempts = 3
+
+// Message is the format we use to tell the MetaFetcher to download
+// the metadata for a given skylink and then add its size to the used space of
+// a given user.
+type Message struct {
+	UserID    primitive.ObjectID
+	SkylinkID primitive.ObjectID
+	Attempts  uint8
+}
+
+// MetaFetcher is a background task that listens for messages on its queue and
+// then processes them.
+type MetaFetcher struct {
+	Queue chan Message
+
+	db     *database.DB
+	portal string
+	logger *logrus.Logger
+}
+
+// New returns a new MetaFetcher instance and starts its internal queue watcher.
+func New(ctx context.Context, db *database.DB, portal string, logger *logrus.Logger) *MetaFetcher {
+	if logger == nil {
+		logger = logrus.New()
+	}
+	mf := MetaFetcher{
+		Queue:  make(chan Message, 1000),
+		db:     db,
+		portal: portal,
+		logger: logger,
+	}
+
+	go mf.threadedStartQueueWatcher(ctx)
+
+	return &mf
+}
+
+// threadedStartQueueWatcher starts a loop over the Queue that processes each
+// incoming message in a separate goroutine.
+func (mf *MetaFetcher) threadedStartQueueWatcher(ctx context.Context) {
+	for m := range mf.Queue {
+		// Process each message in a separate goroutine because fetching the
```

Comment:
This made me think about something, it might be a good idea to verify whether we can squash the message somehow. Because this process takes such a long time, I think a decent portion of the queue will be redundant messages for the same Skylink. We could keep a limited sized cache, or do a database query before executing the actual HEAD call perhaps?

Reply:
Oh I guess we have the added complexity the user's storage needs to be updated. That's not ideal by the way, the metadata fetcher should not be responsible for updating that it feels like to me.

Reply:
1. You are right, we can check the DB before making an expensive HTTP call. I'll add that. This doesn't get in the way of the rest of the logic.
2. I can agree with the idea that the fetcher shouldn't be updating the uploaded amount of the user but this would mean that the user can go *way* over their upload limit before we start limiting them. I'm not sure this is acceptable.